Tests/UnitTests/AdyenDropIn/ComponentContainer/ComponentContainerRouterTests.swift

```diff
@@ -0,0 +1,158 @@
+        }
+
+        override func dismiss(animated: Bool, completion: (() -> Void)? = nil) {
+            dismissCalled = true
+            dismissCompletion = completion
+            completion?()
+        }
+    }
+
+    // MARK: - Helpers
+
+    private func setupSUT() async -> (
+        sut: ComponentContainerRouter,
+        viewControllerSpy: ViewControllerSpy,
+        listenerMock: ComponentContainerRouterListenerMock
+    ) {
+        let viewModelMock = ComponentContainerViewModelProtocolMock()
+        viewModelMock.componentViewController = UIViewController()
+
+        let viewControllerSpy = ViewControllerSpy(viewModel: viewModelMock)
+        let listenerMock = ComponentContainerRouterListenerMock()
+        let sut = ComponentContainerRouter(
+            viewController: viewControllerSpy,
+            listener: listenerMock
+        )
+
+        return (sut, viewControllerSpy, listenerMock)
+    }
+
+    private func makePaymentComponent() async -> PresentableComponentMock {
+        let viewController = UIViewController()
+        let cardPaymentMethodMock = CardPaymentMethodMock(
+            type: .scheme,
+            name: "Card",
+            brands: [.visa, .masterCard]
+        )
+
+        return PresentableComponentMock(
+            paymentMethod: cardPaymentMethodMock,
+            viewController: viewController
+        )
+    }
+
+    private func makeActionComponent() async -> PresentableComponent {
+        let context = AdyenContext(
+            apiContext: Dummy.apiContext,
+            payment: nil,
+            amount: .init(value: 100, currencyCode: "EUR")
+        )
+        let redirect = RedirectComponent(context: context)
+        let viewController = UIViewController()
+        return PresentableComponentWrapper(component: redirect, viewController: viewController)
+    }
+
+    // MARK: - Tests
+
+    @Test
+    func presentPaymentComponent_shouldPushViewController() async throws {
+        // Given
+        let (sut, viewControllerSpy, _) = await setupSUT()
+        let paymentComponent = await makePaymentComponent()
+
+        let navController = UINavigationController(rootViewController: viewControllerSpy)
+        viewControllerSpy.attachNavigationController(navController)
+
+        // When
+        sut.present(paymentComponent: paymentComponent)
+
+        // Then
+        #expect(navController.viewControllers.contains(paymentComponent.viewController))
+    }
+
+    @Test
+    func presentActionComponent_shouldPresentModallyViewController() async throws {
+        // Given
+        let (sut, viewControllerSpy, _) = await setupSUT()
+        let actionComponent = await makeActionComponent()
+
+        // When
+        sut.present(actionComponent: actionComponent, onCancel: nil)
+
+        // Then
+        #expect(viewControllerSpy.presentedViewControllerCaptured != nil)
+    }
+
+    @Test
+    func presentActionComponent_shouldInjectOnCancelCallbackIntoActionWrapper() async throws {
```

Comment:
The test itself is super clear, but now I start to think that the logic itself is complex :)